lib/Modules/ComputeHooks/ComputeHooks.psm1

```diff
@@ -1091,292 +544,461 @@ function Enable-Service {
+        }
+    }
+    if ($DISTRO_URLS[$version]['cluster'] -and $enableCluster) {
+        $ctxt['compute_driver'] = 'hyperv.nova.cluster.driver.HyperVClusterDriver'
+    }
 
-    Start-Service "ovsdb-server"
-    Start-Service "ovs-vswitchd"
+    # Try and create the instanced_dir on the clustered storage. We do not bother testing if the
+    # folder is already there before attempting to create it. There is a chance another node will
+    # create the folder between the Test-Path call and the mkdir call. Might as well try, and check
+    # for the existance of the folder if the command errors out.
+    try {
+        if (!(Test-Path $ctxt["instances_path"])) {
+            New-Item -ItemType Directory $ctxt["instances_path"] | Out-Null
+        }
+   } catch [System.IO.IOException] {
+        if (!(Test-Path $ctxt["instances_path"])) {
+            Throw $_
+        }
+    }
+    return $ctxt
 }
 
-function Start-ConfigureNeutronAgent {
-    $services = Get-CharmServices
-    $vmswitch = Juju-GetVMSwitch
-    $net_type = Get-NetType
-
-    if ($net_type -eq "hyperv"){
-        Disable-Service $services["neutron-ovs"]["service"]
-        Stop-Service $services["neutron-ovs"]["service"] -ErrorAction SilentlyContinue
-
-        Disable-OVS
-        Enable-Service $services["neutron"]["service"]
+function Get-NeutronContext {
+    Write-JujuLog "Generating context for Neutron"
 
-        return $services["neutron"]
+    $logdir = (Get-JujuCharmConfig -Scope 'log-dir').Replace('/', '\')
+    $switchName = Get-JujuVMSwitch
+    if (!(Test-Path $logdir)) {
+        New-Item -ItemType Directory $logdir
     }
 
-    Check-OVSPrerequisites
+    $required = @{
+        "service_protocol"=$null;
+        "service_port"=$null;
+        "auth_host"=$null;
+        "auth_port"=$null;
+        "auth_protocol"=$null;
+        "service_tenant_name"=$null;
+        "service_username"=$null;
+        "service_password"=$null;
+    }
 
-    Disable-Service $services["neutron"]["service"]
-    Stop-Service $services["neutron"]["service"]
+    $optionalCtx = @{
+        "neutron_url"=$null;
+        "quantum_url"=$null;
+        "api_version"=$null;
+    }
 
-    Enable-OVS
-    Enable-Service $services["neutron-ovs"]["service"]
+    $ctx = Get-JujuRelationContext -Relation 'cloud-compute' -RequiredContext $required -OptionalContext $optionalCtx
 
-    Ensure-InternalOVSInterfaces
-    return $services["neutron-ovs"]
-}
+    if (!$ctx.Count -or (!$ctx["neutron_url"] -and !$ctx["quantum_url"])) {
+        Write-JujuWarning "Missing required relation settings for Neutron. Peer not ready?"
+        return @{}
+    }
 
-function Restart-Neutron {
-    $svc = Start-ConfigureNeutronAgent
-    Stop-Service $svc.service
-    Start-Service $svc.service
-}
+    if (!$ctx["neutron_url"]) {
+        $ctx["neutron_url"] = $ctx["quantum_url"]
+    }
+    if (!$ctx["api_version"] -or $ctx["api_version"] -eq 2) {
```

Comment:
Where is this being used in the templates? Should we not set keystone V2/V3 blocks according to the API version?